File: WeeklyCode/week1/Search in Rotated Sorted Array II/gaoqiang.java
public class Solution {
    private int target;

	public boolean search(int[] A, int target) {
		if (A == null || A.length == 0) {
			return false;
		}
		this.target = target;
		return binarySearch(A, 0, A.length - 1);
	}

	private boolean binarySearch(int[] A, int start, int end) {
		if (start > end)
			return false;
		int mid = (start + end) / 2;
		if (A[mid] == target) {
			return true;
		}
		if (A[start] == A[mid] && A[start] == A[end]) {
			for (int i = start; i < mid; i++) {
				if (A[i] != A[start]) {
					return binarySearch(A, start, mid - 1);
				}
			}
			return binarySearch(A, mid + 1, end);
		}
		if (A[start] <= A[mid]) {
			if (A[start] <= target && A[mid] > target) {
				return binarySearch(A, start, mid - 1);
			} else {
				return binarySearch(A, mid + 1, end);
			}
		} else {
			if (A[mid] < target && A[end] >= target) {
				return binarySearch(A, mid + 1, end);
			} else {
				return binarySearch(A, start, mid - 1);
			}
		}
	}

}

File: WeeklyCode/week6/Search for a Range/gaoqiang.py
class Solution:
    # @param A, a list of integers
    # @param target, an integer to be searched
    # @return a list of length 2, [index1, index2]
    def searchRange(self, A, target):
    	left = self.binarySearchLeft(A, 0, len(A)-1, target)
    	right = self.binarySearchRight(A, 0, len(A)-1, target)
    	return [left,right]

    def binarySearchLeft(self, A, start, end, target):
    	result = self.binarySearch(A, start, end, target)
    	while result!=0 and result!=-1 and A[result-1]==target:
    		result = self.binarySearchLeft(A, start, result-1, target)
    	return result

    def binarySearchRight(self, A, start, end, target):
    	result = self.binarySearch(A, start, end, target)
    	while result!=len(A)-1 and result!=-1 and A[result+1]==target:
    		result = self.binarySearchRight(A, result+1, end, target)
    	return result

    def binarySearch(self, A, start, end, target):
    	if start > end:
    		return -1
    	mid = (start+end)/2
    	if A[mid]==target:
    		return mid
    	if A[mid]<target:
    		return self.binarySearch(A, mid+1, end, target)
    	return self.binarySearch(A, start, mid-1, target)


if __name__ == '__main__':
	s = Solution()
	print s.searchRange([],11)